src/core/control/CrashHandler.cpp

```diff
@@ -1,27 +1,83 @@
 #include "CrashHandler.h"
 
-#include <string>  // for string
+#include <array>        // for std::array
+#include <atomic>       // for std::atomic
+#include <csignal>      // for std::signal
+#include <fstream>      // for std::ofstream
+#include <string>       // for std::string
+#include <string_view>  // for std::string_view
 
-#include <glib.h>  // for g_warning, g_error
+#include <fbbe/stacktrace.h>  // for fbbe::stacktrace
+#include <glib.h>             // for g_warning, g_error
 
 #include "control/xojfile/SaveHandler.h"  // for SaveHandler
-#include "util/PathUtil.h"                // for getConfigFile
-#include "util/i18n.h"                    // for FC, _F, _
+#include "util/PathUtil.h"
+#include "util/PathUtil.h"    // for getConfigFile
+#include "util/Stacktrace.h"  // for Stacktrace::printStracktrace
+#include "util/i18n.h"        // for FC, _F, _
 
+#include "config-dev.h"
 #include "filesystem.h"  // for path
 
-static Document* document = nullptr;
+
+static std::atomic<Document*> document = nullptr;
+static std::atomic<bool> alreadyCrashed = false;
 
 void setEmergencyDocument(Document* doc) { document = doc; }
 
-#ifdef _WIN32
-#include "CrashHandlerWindows.h"
-#else
-#include "CrashHandlerUnix.h"
-#endif
+/**
+ * Print crash log to config directory
+ */
+static void crashHandler(int sig) {
+    if (alreadyCrashed) {  // crasehd again on emergency save
+        exit(2);
+    }
+    alreadyCrashed = true;
+
+    g_warning("[Crash Handler] Crashed with signal %i", sig);
+
+    auto curtime = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+
+    std::array<char, 128> stime;
+    size_t size = 0;
+    {
+        static constexpr std::array<char, 10> error_log_name{"errorlog."};
+        std::memcpy(stime.data(), error_log_name.data(), error_log_name.size() - 1);
+        assert(stime[8] == '.');
+        size = strftime(&stime[9], stime.size() - error_log_name.size(), "%Y%m%d-%H%M%S", localtime(&curtime));
+        std::memcpy(stime.data() + size + 9, ".log", 4);
+    }
+    auto const& errorlogPath = Util::getCacheSubfolder(ERRORLOG_DIR) / std::string_view(stime.data(), size + 9 + 4);
+    std::ofstream fp(errorlogPath);
+    if (fp) {}
+
+    fp << FORMAT_STR("Date: {1}") % ctime(&curtime);
+    fp << FORMAT_STR("Error: signal {1}") % sig;
+    fp << "\n";
+    Stacktrace::printStracktrace(std::cerr);
+    Stacktrace::printStracktrace(fp);
+    if (fp) {
+        g_warning("[Crash Handler] Wrote crash log to: %s", errorlogPath.c_str());
+        fp.close();
+    }
+
+    emergencySave();
+
+    exit(1);
+}
+
+void installCrashHandlers() {
+    std::signal(SIGTERM, crashHandler);
+    std::signal(SIGINT, crashHandler);
```

Comment:
`SIGTERM` and `SIGINT` were not handled as crashes in the previous version, is this change ok?

Reply:
Actually for sigint no, sigint does not indicate, that the process is in an undetermined/ broken state. So we would require finally a signal handler, which only sets a flag. That flag must be checked to then open a close request.

For the terminate signal I would say yes but that's pretty irrelevant. Mostly any other signal is raised before a thread terminates.

Reply:
I did a few experiments and `Xournal++ (1.1.1)` just closes on `sigint` and `sigterm`.

In those cases reporting a crash it may be weird but saving an emergency file it is definitively handy.

It may happen that someone presses `ctrl-c` in the incorrect terminal by mistake and loses all the modifications. Having an emergency doc it will mitigate the damage.

Reply:
See also https://github.com/xournalpp/xournalpp/issues/1083#issuecomment-485690664